### PR TITLE
[FLINK-27399][Connector/Pulsar] Change initial consuming position setting logic for better handle the checkpoint.

### PR DESCRIPTION
**Pull request by syhily**
## Brief change log

This task modifies the `flink-connector-pulsar` module, adds some new mechanisms to set the initial consuming position.

- Change `StartCursor`, add new useful methods, and deprecate the confused `fromMessageTime()` method.
- Change `StopCursor`, and add new useful methods.
- Introduce a new `SplitAssigner` for assigning the splits among the Pulsar readers. Make the partition assignment logic clear and testable.
- Change the start position seeking mechanism from Pulsar consumer API to Pulsar admin API. Don't reset the start position when the topic has a subscription.

## Verifying this change

This change is already covered by existing tests, such as:

- `PulsarSourceITCase`
- `PulsarSourceEnumeratorTest`
- `PulsarOrderedPartitionSplitReaderTest`

We add new tests for new partition assign logic:

- `NormalSplitAssignerTest`
- `SharedSplitAssignerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduces a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-python/pyflink/datastream/connectors/pulsar.py
Comment (deadwind4) [inline flink-python/pyflink/datastream/connectors/pulsar.py:141]: Should we like this to keep forward-compatibility?

https://github.com/apache/flink/blob/314e276f6c6bff990e82515d6ce90fd6a7c9561d/flink-python/pyflink/datastream/connectors/pulsar.py#L192-L207

Comment (syhily) [reply flink-python/pyflink/datastream/connectors/pulsar.py:141]: We have changed the MessgeTime to PublishTime in Java API. This should be a mistake in both Java and Python implementation. So I just introduce this breaking change.

Comment (syhily) [reply flink-python/pyflink/datastream/connectors/pulsar.py]: Add this breaking change may not sound like a good choice. I add a new method and deprecated the old ones.

File: docs/content.zh/docs/connectors/datastream/pulsar.md
Comment (imaffe) [inline docs/content.zh/docs/connectors/datastream/pulsar.md:453]: Nice renaming~

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java
Comment (imaffe) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:244]: When will individual reader receive no more splits ?

**Review by imaffe — COMMENTED** [review]
I reviewed until the NormlSplitAssigner. Will continue review tomorrow.

**Comment by MartijnVisser** [comment]
@PatrickRen Can you also have a look at this PR? I've understood that this PR should help resolve this blocker test stability [FLINK-26721](https://issues.apache.org/jira/browse/FLINK-26721)

**Comment by tisonkun** [comment]
@MartijnVisser Thank you. I'll review the patch in this week. Actually I ever try to request myself as a reviewer but forget several times >_<

**Review by tisonkun — COMMENTED** [review]
Generally looks good. The effective changes are:

1. Changing `SplitAssignmentState` to `SplitAssigner`
2. Create subscription via PulsarAdmin API instead of consumer API and seek.

Comments inline.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/start/MessageIdStartCursor.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/start/MessageIdStartCursor.java:56]: Yep. We just reuse the same logic here in favor of DRY.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/StopCursor.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/StopCursor.java:111]: Yep.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/start/MessageIdStartCursor.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/start/MessageIdStartCursor.java:51]: Yep.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/CursorPosition.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/CursorPosition.java:52]: Yep.

**Comment by MartijnVisser** [comment]
@syhily Can you resolve the latest review comments?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java
Comment (imaffe) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java:116]: The pendingSplits contains 1. added back splits due to reader crash 2. possible new splits due to newly subscribed topics. Is this correct ?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java
Comment (imaffe) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java:55]: nit: NormalSplitsAssigner would let me think the others are "abnormal", I would suggest "ExclusiveAndFailoverSplitAssigner" but this name is too long, not desired as well. Any thoughts here ?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java
Comment (imaffe) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java:39]: what is our current status for KEY_SHARED subscriptionType ?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java
Comment (imaffe) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java:40]: Shouldn't KEY_SHARED use this splitAssigner as well ?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/StopCursor.java
Comment (imaffe) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/StopCursor.java:111]: To me,at vs after is similar to how exclusive field. What would be the semantics if we add the exclusive field ?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java:55]: Cool. This would be better.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java:39]: They are using `NormalSplitAssigner.java`.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/SharedSplitAssigner.java:116]: Yep

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/assigner/NormalSplitAssigner.java]: Yep

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/MessageIdStopCursor.java
Comment (tisonkun) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/MessageIdStopCursor.java:58]: It seems we somehow revert the boolean logic here?

**Review by tisonkun — COMMENTED** [review]
Generally looks good. One comment inline. You may give another pass to all interval inclusive/exclusive conditions.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/MessageIdStopCursor.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/MessageIdStopCursor.java:58]: Nope, this is a wrong use in `PulsarPartitionSplitReaderBase`, we just fix the bugs.

**Review by tisonkun — APPROVED** [review]
Thanks for your contribution @syhily! LGTM.

Merging if CI gives green...

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/LatestMessageStopCursor.java
Comment (tisonkun) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/cursor/stop/LatestMessageStopCursor.java:32]: I'd prefer to `implements StopCursor` and keep `MessageIdStopCursor` holds a final `messageId`.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java
Comment (MartijnVisser) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:39]: Will you create a follow-up Jira ticket to address this issue? We should not disable a test without any reference on why and what.

Comment (tisonkun) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:39]: +1 This is the heaviest and most unstable test of Pulsar connector.

If you simply disable it, you'd better do it in a separate patch with a reference. Otherwise, I'm unconfident to other changes in this patch.

**Review by tisonkun — CHANGES_REQUESTED** [review]

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java
Comment (rmetzger) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:39]: Confident or unconfident? I think you approved this pull request earlier?

> you'd better do it in a separate patch with a reference
I agree. It would be nice to have a JIRA ticket to re-enable this, mention the JIRA in the `@Disabled` annotation and do this in a separate commit (but it can be in this PR)

Comment (tisonkun) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:39]: > unconfident

There're several follow-up commits including tests disabled. I don't think we should mix up a test fix with test disable.

It seems the current status stuck due to failure cannot reproduce locally. I'll try to help on this point this week.

Comment (rmetzger) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:39]: Thanks for clarifying. I didn't check the commits after your approval in detail.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java
Comment (tisonkun) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java:102]: Cool.

It seems CI failed on Kafka tests. But you may retrigger it and we can merge this one once CI gives green.

**Comment by syhily** [comment]
@tisonkun Finally, the ci turns green.

**Review by tisonkun — APPROVED** [review]

**Pull request merged**
